File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            if (nums[mid] == target) {
                return mid;
            }
            
            // If the left half is sorted
            if (nums[left] <= nums[mid]) {
                if (target >= nums[left] && target < nums[mid]) {
                    right = mid - 1;  // target is in the left half
                } else {
                    left = mid + 1;   // target is in the right half
                }
            }
            // If the right half is sorted
            else {
                if (target > nums[mid] && target <= nums[right]) {
                    left = mid + 1;   // target is in the right half
                } else {
                    right = mid - 1;  // target is in the left half
                }
            }
        }
        
        return -1;  // target was not found
    }
}

File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1

        while left <= right:
            mid = (left + right) // 2

            if (nums[mid] == target):
                return mid

            if (nums[left] <= nums[mid]) :
                if (target >= nums[left] and nums[mid] >= target):
                    right = mid - 1
                else:
                    left = mid + 1
            else:
                if (target <= nums[right] and nums[mid] <= target):
                    left = mid + 1
                else:
                    right = mid - 1
        return -1

